### DISCUSSION TITLE
How to declare classes for use in RHS?

### DISCUSSION THREAD

**Discussion by jstandard** [evrete/evrete#57] [General]
Hello,

I'm planning to migrate a project from Jess to Evrete.  I've been using the Evrete API to get a feel for how it works.  However, I've found an issue I can't seem to fix.  Below is a simple example where if an Order is greater than $100 we provide a 10% Coupon off the next Order.  When I run this code, I get the message: 

**May 25, 2023 6:35:27 PM org.evrete.runtime.AbstractRuleSession insertAtomic
WARNING: Can not map type for 'example.rule.Coupon', insert operation skipped.**

If I add the Coupon.class to the forEach method 
```
.forEach("$order", Order.class "$coupon", Coupon.class)
```
the problem goes away.  However, the rule doesn't fire any longer.

Is there a way to declare classes for use in the rules engine, if it's not part of the LHS?

```
public class ExampleEvreteTest {
    private KnowledgeService service = new KnowledgeService();
    private Knowledge knowledge = service.newKnowledge();
    private StatefulSession session;

    @Test
    public void testCouponAdded() {
        knowledge.newRule("declare rule")
                .forEach("$order", Order.class)
                .where("$order.total > 100")
                .execute(ctx -> ctx.insert(new Coupon("10% off next order!")));

        session = knowledge.newStatefulSession();
        session.insert(new Order(99));
        session.insert(new Order(100));
        session.insert(new Order(101));
        session.fire();

        List<Coupon> coupons = new ArrayList<>();
        Consumer<Object> couponCheck = fact -> {
            if (fact instanceof Coupon) {
                coupons.add((Coupon) fact);
            }
        };
        session.forEachFact(couponCheck);
        assertTrue(!coupons.isEmpty() && coupons.size() == 1);
    }

}
```

**Comment by andbi** [comment]
Hi! In your case the engine has detected that you're inserting a fact which type is never used in LHS, not once. If you continue migration, and add rules dealing with the Order type, then the issue will go away.

**Reply by jstandard** [reply]
Sorry, maybe I wasn't clear.  The problem is inserting a new Coupon object, if there isn't a rule that deals with a Coupon.  I just want to insert Coupons into working memory, when the rule fires.  Maybe I'm thinking about this problem in the wrong way?

In Jess, you can declare objects that don't appear as facts.  Therefore, they are known to the system when it's time to insert them when the rule fires.

**Reply by andbi** [reply]
So that `Coupon` object is just a helper object that won't be a part of any rule by design, but you need to be able to access it somewhere in the future. Do I get it right?

Evrete won't insert it into the RETE memory but provides means of storing such auxiliary objects via the `org.evrete.api.Environment` interface, which basically is a map with convenient type casting accessors. Both `Knowledge` and `Session` implement this API.

https://github.com/evrete/evrete/blob/master/evrete-core/src/main/java/org/evrete/api/Environment.java

You can access your session's environment from `RhsContext` by invoking the `ctx.getRuntime()` method.

I will update the API docs in the next release, but for now, it's essential to understand the inheritance model of Environment instances:
If you store an object inside a Knowledge instance, it will be populated in every session spawned from that Knowledge instance.

**Reply by jstandard** [reply]
Thanks.  Well this is a contrived example, but yes there are helper objects that aren't part of the rule itself.  And they need to be inserted in working memory in the future.

I'll look at the Environment implementation within the source and get familiar with it.  I appreciate your prompt responses!
